File: backend/modules/CompanyProfile/CromaticaCreativa.Modules.CompanyProfile.Infrastructure/Persistence/Mappers/CompanyProfileMapper.cs
using CromaticaCreativa.Modules.CompanyProfile.Domain.Aggregates;
using CromaticaCreativa.Modules.CompanyProfile.Domain.ValueObjects;
using CromaticaCreativa.Modules.CompanyProfile.Infrastructure.Persistence.Models;

namespace CromaticaCreativa.Modules.CompanyProfile.Infrastructure.Persistence.Mappers;

public static class CompanyProfileMapper
{
    public static CompanyContactInformation ToDomain(CompanyProfileModel model)
    {
        ArgumentNullException.ThrowIfNull(model);

        ValidateEmailTypes(model);
        var recipients = model.Emails
            .Where(email => email.Type == CompanyProfilePersistenceValues.ContactRequestRecipient)
            .ToArray();

        if (recipients.Length != 1)
        {
            throw new InvalidOperationException(
                $"Company profile '{model.Id}' must contain exactly one contact-request recipient email row.");
        }

        var profile = CompanyContactInformation.Create(
            new CompanyContactInformationId(model.Id),
            new EmailAddress(recipients[0].Address));

        var seenPhoneTypes = new HashSet<string>(StringComparer.Ordinal);
        foreach (var phone in model.Phones)
        {
            if (!seenPhoneTypes.Add(phone.Type))
            {
                throw new InvalidOperationException(
                    $"Company profile '{model.Id}' contains more than one '{phone.Type}' phone row.");
            }

            switch (phone.Type)
            {
                case CompanyProfilePersistenceValues.Phone:
                    profile.ChangePhone(new PhoneNumber(phone.Number));
                    break;
                case CompanyProfilePersistenceValues.WhatsApp:
                    profile.ChangeWhatsApp(new PhoneNumber(phone.Number));
                    break;
                default:
                    throw new InvalidOperationException(
                        $"Company profile '{model.Id}' has unsupported phone type '{phone.Type}'.");
            }
        }

        var publicEmails = model.Emails
            .Where(email => email.Type == CompanyProfilePersistenceValues.PublicEmail)
            .ToArray();
        if (publicEmails.Length > 1)
        {
            throw new InvalidOperationException(
                $"Company profile '{model.Id}' contains more than one public email row.");
        }

        if (publicEmails.Length == 1)
        {
            profile.ChangePublicEmail(new EmailAddress(publicEmails[0].Address));
        }

        if (model.Location is not null)
        {
            GeoCoordinates? coordinates = model.Location.Latitude is null
                ? null
                : new GeoCoordinates(model.Location.Latitude.Value, model.Location.Longitude
                    ?? throw new InvalidOperationException(
                        $"Location '{model.Location.Id}' has latitude without longitude."));

            if (model.Location.Latitude is null && model.Location.Longitude is not null)
            {
                throw new InvalidOperationException(
                    $"Location '{model.Location.Id}' has longitude without latitude.");
            }

            profile.SetLocation(
                new CompanyLocationId(model.Location.Id),
                new Address(model.Location.Address),
                coordinates);
        }

        var seenNetworks = new HashSet<string>(StringComparer.Ordinal);
        foreach (var socialLink in model.SocialLinks.OrderBy(link => link.DisplayOrder).ThenBy(link => link.Id))
        {
            if (!seenNetworks.Add(socialLink.Network))
            {
                throw new InvalidOperationException(
                    $"Company profile '{model.Id}' contains more than one social link for '{socialLink.Network}'.");
            }

            var added = profile.AddSocialLink(new SocialLink(
                socialLink.Network,
                new ExternalUrl(socialLink.Url)));

            if (!added)
            {
                throw new InvalidOperationException(
                    $"Company profile '{model.Id}' contains duplicate social link data for '{socialLink.Network}'.");
            }
        }

        return profile;
    }

    public static CompanyProfileModel ToPersistence(
        CompanyContactInformation profile,
        CompanyProfileModel? existingModel = null)
    {
        ArgumentNullException.ThrowIfNull(profile);

        if (existingModel is not null && existingModel.Id != profile.Id.Value)
        {
            throw new ArgumentException(
                "The existing persistence model belongs to a different company profile.",
                nameof(existingModel));
        }

        var model = new CompanyProfileModel
        {
            Id = profile.Id.Value,
            SingletonKey = 1
        };

        if (profile.Phone is not null)
        {
            model.Phones.Add(CreatePhoneModel(
                profile.Id.Value,
                CompanyProfilePersistenceValues.Phone,
                profile.Phone.Value,
                existingModel));
        }

        if (profile.WhatsApp is not null)
        {
            model.Phones.Add(CreatePhoneModel(
                profile.Id.Value,
                CompanyProfilePersistenceValues.WhatsApp,
                profile.WhatsApp.Value,
                existingModel));
        }

        if (profile.PublicEmail is not null)
        {
            model.Emails.Add(CreateEmailModel(
                profile.Id.Value,
                CompanyProfilePersistenceValues.PublicEmail,
                profile.PublicEmail.Value,
                existingModel));
        }

        model.Emails.Add(CreateEmailModel(
            profile.Id.Value,
            CompanyProfilePersistenceValues.ContactRequestRecipient,
            profile.ContactRequestRecipientEmail.Value,
            existingModel));

        if (profile.Location is not null)
        {
            model.Location = new LocationModel
            {
                Id = profile.Location.Id.Value,
                CompanyProfileId = profile.Id.Value,
                Address = profile.Location.Address.Value,
                Latitude = profile.Location.Coordinates?.Latitude,
                Longitude = profile.Location.Coordinates?.Longitude
            };
        }

        var displayOrder = 0;
        foreach (var socialLink in profile.SocialLinks)
        {
            var existingLink = existingModel?.SocialLinks.SingleOrDefault(
                item => string.Equals(item.Network, socialLink.Network, StringComparison.Ordinal));

            model.SocialLinks.Add(new SocialLinkModel
            {
                Id = existingLink?.Id ?? Guid.NewGuid(),
                CompanyProfileId = profile.Id.Value,
                Network = socialLink.Network,
                Url = socialLink.Url.Value,
                DisplayOrder = displayOrder++
            });
        }

        return model;
    }

    private static void ValidateEmailTypes(CompanyProfileModel model)
    {
        foreach (var email in model.Emails)
        {
            if (email.Type is not CompanyProfilePersistenceValues.PublicEmail
                and not CompanyProfilePersistenceValues.ContactRequestRecipient)
            {
                throw new InvalidOperationException(
                    $"Company profile '{model.Id}' has unsupported email type '{email.Type}'.");
            }
        }
    }

    private static PhoneModel CreatePhoneModel(
        Guid companyProfileId,
        string type,
        string number,
        CompanyProfileModel? existingModel)
    {
        var existingPhone = existingModel?.Phones.SingleOrDefault(item => item.Type == type);
        return new PhoneModel
        {
            Id = existingPhone?.Id ?? Guid.NewGuid(),
            CompanyProfileId = companyProfileId,
            Type = type,
            Number = number
        };
    }

    private static EmailModel CreateEmailModel(
        Guid companyProfileId,
        string type,
        string address,
        CompanyProfileModel? existingModel)
    {
        var existingEmail = existingModel?.Emails.SingleOrDefault(item => item.Type == type);
        return new EmailModel
        {
            Id = existingEmail?.Id ?? Guid.NewGuid(),
            CompanyProfileId = companyProfileId,
            Type = type,
            Address = address
        };
    }
}

File: backend/src/modules/CompanyProfile/CompanyProfile.Infrastructure/Persistence/Mappers/CompanyProfileMapper.ts
import { randomUUID } from 'node:crypto';
import { CompanyProfilePersistenceDto } from '../../../CompanyProfile.Commons/DTOs/CompanyProfilePersistenceDto';
import { CompanyContactInformation } from '../../../CompanyProfile.Domain/Aggregates/CompanyContactInformation';
import { Address } from '../../../CompanyProfile.Domain/ValueObjects/Address';
import { CompanyContactInformationId } from '../../../CompanyProfile.Domain/ValueObjects/CompanyContactInformationId';
import { CompanyLocation } from '../../../CompanyProfile.Domain/ValueObjects/CompanyLocation';
import { EmailAddress } from '../../../CompanyProfile.Domain/ValueObjects/EmailAddress';
import { ExternalUrl } from '../../../CompanyProfile.Domain/ValueObjects/ExternalUrl';
import { GeoCoordinates } from '../../../CompanyProfile.Domain/ValueObjects/GeoCoordinates';
import { PhoneNumber } from '../../../CompanyProfile.Domain/ValueObjects/PhoneNumber';
import { SocialLink } from '../../../CompanyProfile.Domain/ValueObjects/SocialLink';
import { CompanyProfilePersistenceModel } from '../Models/CompanyProfilePersistenceModel';
import { EmailPersistenceModel } from '../Models/EmailPersistenceModel';
import { LocationPersistenceModel } from '../Models/LocationPersistenceModel';
import { PhonePersistenceModel } from '../Models/PhonePersistenceModel';
import { SocialLinkPersistenceModel } from '../Models/SocialLinkPersistenceModel';

export class CompanyProfileMapper {
  public static toDomain(model: CompanyProfilePersistenceDto): CompanyContactInformation {
    const profile = CompanyContactInformation.create(
      new CompanyContactInformationId(model.id),
      new EmailAddress(model.contactRequestRecipientEmail),
    );

    for (const item of this.ordered(model.phones ?? [])) {
      if (!profile.addPhone(new PhoneNumber(item.number))) {
        throw new Error(`El perfil '${model.id}' contiene teléfonos públicos duplicados.`);
      }
    }

    for (const item of this.ordered(model.emails ?? [])) {
      if (!profile.addEmail(new EmailAddress(item.address))) {
        throw new Error(`El perfil '${model.id}' contiene correos públicos duplicados.`);
      }
    }

    for (const item of this.ordered(model.socialLinks ?? [])) {
      if (!profile.addSocialLink(new SocialLink(item.network, new ExternalUrl(item.url)))) {
        throw new Error(`El perfil '${model.id}' contiene más de un enlace para la red '${item.network}'.`);
      }
    }

    if (model.location) {
      profile.setLocation(new CompanyLocation(
        new Address(model.location.address),
        new GeoCoordinates(model.location.latitude, model.location.longitude),
      ));
    }

    return profile;
  }

  public static toPersistence(
    profile: CompanyContactInformation,
    existing: CompanyProfilePersistenceModel | null = null,
  ): CompanyProfilePersistenceModel {
    if (existing && existing.id !== profile.id.value) {
      throw new Error('El estado persistido existente pertenece a otro perfil de empresa.');
    }

    const model = new CompanyProfilePersistenceModel();
    model.id = profile.id.value;
    model.singletonKey = 1;
    model.contactRequestRecipientEmail = profile.contactRequestRecipientEmail.value;
    model.phones = profile.phones.map((phone, displayOrder) => Object.assign(new PhonePersistenceModel(), {
      id: existing?.phones?.find((item) => item.number === phone.value)?.id ?? randomUUID(),
      companyProfileId: profile.id.value,
      number: phone.value,
      displayOrder,
      companyProfile: model,
    }));
    model.emails = profile.emails.map((email, displayOrder) => Object.assign(new EmailPersistenceModel(), {
      id: existing?.emails?.find((item) => item.address === email.value)?.id ?? randomUUID(),
      companyProfileId: profile.id.value,
      address: email.value,
      displayOrder,
      companyProfile: model,
    }));
    model.socialLinks = profile.socialLinks.map((link, displayOrder) => Object.assign(new SocialLinkPersistenceModel(), {
      id: existing?.socialLinks?.find((item) => this.sameNetwork(item.network, link.network))?.id ?? randomUUID(),
      companyProfileId: profile.id.value,
      network: link.network,
      url: link.url.value,
      displayOrder,
      companyProfile: model,
    }));
    model.location = profile.location
      ? Object.assign(new LocationPersistenceModel(), {
        companyProfileId: profile.id.value,
        address: profile.location.address.value,
        latitude: profile.location.coordinates.latitude,
        longitude: profile.location.coordinates.longitude,
        companyProfile: model,
      })
      : null;
    return model;
  }

  private static ordered<T extends { id: string; displayOrder: number }>(items: readonly T[]): T[] {
    if (items.some((item) => !Number.isInteger(item.displayOrder) || item.displayOrder < 0)) {
      throw new Error('El estado persistido contiene un orden de visualización inválido.');
    }
    return [...items].sort((left, right) => left.displayOrder - right.displayOrder || left.id.localeCompare(right.id));
  }

  private static sameNetwork(left: string, right: string): boolean {
    return left.trim().toLocaleLowerCase('en-US') === right.trim().toLocaleLowerCase('en-US');
  }
}
